src/main/scala/scalaz/analytics/AnalyticsModule.scala

```diff
@@ -100,19 +100,33 @@ trait AnalyticsModule {
    * The DataSet/DataStream operations of scalaz-analytics
    */
   trait Ops[F[_]] {
-    // Unbounded
+    // Unbounded dataset ops
+    // (These ops require only bounded space, even with unbounded data)
     def map[A, B](ds: F[A])(f: A =>: B): F[B]
+    def flatMap[A, B](ds: F[A])(f: A =>: DataSet[B]): F[B]
     def filter[A](ds: F[A])(f: A =>: Boolean): F[A]
+    // Streaming aggregations
+    // (These ops produce cumulative results, some may require unbounded space)
+    def scan[A, B](ds: F[A])(initial: Unit =>: B)(f: (B, A) =>: B): F[B]
 
-    // Bounded
-    def fold[A, B](ds: F[A])(window: Window)(initial: A =>: B)(f: (B, A) =>: B): F[B]
+    def scanAggregateBy[A, K, V](ds: F[A])(g: A =>: K)(initial: Unit =>: V)(
+      f: (V, A) =>: V
+    ): F[(K, V)]
+
+    // Bounded dataset ops
+    // (These ops can only be performed on bounded subsets of data)
+    def fold[A, B](ds: F[A])(window: Window)(initial: Unit =>: B)(f: (B, A) =>: B): F[B]
+
+    def aggregateBy[A, K, V](ds: F[A])(window: Window)(g: A =>: K)(initial: Unit =>: V)(
```

Comment:
What is this exactly? Can it be broken into more fundamental pieces? GroupBy + Fold or something perhaps?

Reply:
Note: I modeled grouping+aggregation this way on the basis of comments [like this one](https://github.com/scalaz/scalaz-analytics/issues/4#issuecomment-416032416) in the dataset design discussion (#4) that questioned the utility of separating grouping and aggregation.

The other day, I toyed with the idea of implementing the `aggregateBy` operation in terms of separate `groupBy` and `fold` operations:
```scala
def groupBy[A, K](ds: F[A])(g: A =>: K): F[Grouped[K, A]]

def fold[A, B](ds: F[A])(window: Window)(initial: Unit =>: B)(f: (B, A) =>: B): F[B]
```
This doesn't quite work:
```scala
def aggregateBy[A, K, V](ds: F[A])(window: Window)(g: A =>: K)
                        (initial: Unit =>: V)(f: (V, A) =>: V): F[(K, V)] =
  fold(groupBy(ds)(g))(window)(initial)(f)
```
(The types don't quite line up, because you want the fold to result in an `F[(K, V)]` but the initial value for the fold is a `V` and not a `(K, V)`.)

Instead, you can introduce a separate `groupFold` operation:
```scala
def groupFold[A, K, V](ds: F[Grouped[K, A]])(window: Window)
                      (initial: Unit =>: V)(f: (V, A) =>: V): F[(K, V)]

def aggregateBy[A, K, V](ds: F[A])(window: Window)(g: A =>: K)
                        (initial: Unit =>: V)(f: (V, A) =>: V): F[(K, V)] =
  groupFold(groupBy(ds)(g))(window)(initial)(f)
```

All that said, it's not obvious to me that there's real value in factoring out the `groupBy`.

Reply:
Another approach that might work, without requiring separate methods like `groupFold`, would be to allow the fold to produce an `F[Grouped[K, V]]` rather than an `F[(K, V)]`:
```scala
def groupBy[A, K](ds: F[A])(g: A =>: K): F[Grouped[K, A]]

def aggregateBy[A, K, V](ds: F[A])(window: Window)(g: A =>: K)
                        (initial: Unit =>: V)(f: (V, A) =>: V): F[Grouped[K, V]] = {
  val initial2: Unit =>: Grouped[K, V] = ??? // derive from `initial`
  val f2: (Grouped[K, V], Grouped[K, A]) =>: Grouped[K, V] = ??? // derive from `f`
  fold(groupBy(ds)(g))(window)(initial2)(f2)
}
```

This approach might make the case for factoring out `groupBy` a bit more compelling, since it can reuse a few aggregation primitives (e.g., `fold`, `scan`) rather than requiring Grouped-specific aggregation operations like `groupedFold`, `groupedScan`, etc.  Even then, however, it's not clear that the added complexity pays for itself.

The tupled representation can be recovered with an `ungroup` operation:
```scala
def ungroup[K, V](ds: F[Grouped[K, V]]): F[(K, V)]

def aggregateBy[A, K, V](ds: F[A])(window: Window)(g: A =>: K)
                        (initial: Unit =>: V)(f: (V, A) =>: V): F[(K, V)] =
  ungroup(fold(groupBy(ds)(g))(window)(initial2)(f2)) 
  // where initial2 and f2 are transformations as above
```